### PR TITLE
feat: Add possibility to record calls

### PR DESCRIPTION
**Pull request by Urogna**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
     }
   }
 
-  const errorMessage = `Request ${req.url} didn't match any registered route.`;
+  const obfuscatedBody = JSON.parse(JSON.stringify(req.body));
+  recordingsContext.deleteBodyAttributes.forEach((attr) => {
+    try {
+      console.log('Deleting attribute', attr);
+      eval(`delete obfuscatedBody.${attr}`);
```
Comment (petermanser) [inline index.js:201]: delete obfuscatedBody[attr];

Comment (petermanser) [inline index.js:201]:
```function deleteNestedProperty(obj: Record<string, any>, path: string): void {
    const keys = path.split('.');
    const lastKey = keys.pop();
    const lastObj = keys.reduce((acc, key) => acc && acc[key], obj);

    if (lastObj && lastKey in lastObj) {
        delete lastObj[lastKey];
    }
}

deleteNestedProperty(obfuscatedBody, attr);

```

File: index.js [hunk 2]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
...
+      console.log('No attribute to delete: ', attr);
+    }
+  });
+  const dataToHash = {
+    url: req.url,
+    body: obfuscatedBody,
+    method: req.method,
+    headers: req.headers,
+  };
+  const crypto = require('crypto');
```
Comment (petermanser) [inline index.js:212]: move to top

File: index.js [hunk 3]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
...
+  const hash = crypto.createHash('sha512').update(JSON.stringify(dataToHash)).digest('hex');
+  if (recordingsContext.active) {
+    try {
+      const host = req.url.split('/')[1];
```
Comment (petermanser) [inline index.js:216]: const [_, host, ...routeParts] = req.url.split('/');
        const route = `/${routeParts.join('/')}`;

File: index.js [hunk 4]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
...
+      const route = req.url.replace(`/${host}`, '');
+      const targetUrl = `https://${host}${route}`;
+      console.log('Proxying from ', req.url, ' to', targetUrl);
+      const proxyRes = await fetch(targetUrl, {
+        method: req.method,
+        headers: { ...req.headers, 'Access-Control-Allow-Origin': '*' },
+        ...(req.method !== 'GET' && req.method !== 'HEAD' ? { body: JSON.stringify(req.body) } : {}),
```
Comment (petermanser) [inline index.js:223]: && req.body

File: index.js [hunk 5]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
...
+  const obfuscatedBody = JSON.parse(JSON.stringify(req.body));
+  recordingsContext.deleteBodyAttributes.forEach((attr) => {
+    try {
+      console.log('Deleting attribute', attr);
+      eval(`delete obfuscatedBody.${attr}`);
+    } catch {
+      console.log('No attribute to delete: ', attr);
+    }
+  });
+  const dataToHash = {
```
Comment (petermanser) [inline index.js:206]: move to own function `getHashForRequest`?

File: index.js (hunk 1 above)
Comment (Urogna) [reply index.js]: 👍

File: index.js (hunk 2 above)
Comment (Urogna) [reply index.js]: 👍

File: index.js [hunk 6]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
     }
   }
 
-  const errorMessage = `Request ${req.url} didn't match any registered route.`;
+  const obfuscatedBody = JSON.parse(JSON.stringify(req.body));
```
[the remark this answers is not in the document]
Comment (Urogna) [reply index.js]: `obfuscatedReqBodyForHash`

File: index.js [hunk 7]
```diff
@@ -168,7 +194,71 @@ app.all('*', (req, res) => {
...
+    } catch (e) {
+      console.log({
+        error: e.message + ' ' + req.method,
+        url: req.url,
+      });
+    }
+    return;
+  }
+
+  const responseFromHash = Object.values(recordings).find((rec) => rec[hash]);
```
[the remark this answers is not in the document]
Comment (Urogna) [reply index.js]: Good catch... this was before I always had namespace context

File: index.js (hunk 3 above)
Comment (Urogna) [reply index.js]: 👍

File: index.js (hunk 4 above)
Comment (Urogna) [reply index.js]: 👍

File: index.js (hunk 5 above)
Comment (Urogna) [reply index.js:221]: 👍

File: index.js [hunk 8]
```diff
@@ -102,6 +110,25 @@ route.post('/reset/calls', (_req, res) => {
   res.sendStatus(204);
 });
 
+route.post('/record', (req, res) => {
```
Comment (nickredmark) [inline index.js:113]: proposal. just one route with two methods:

POST /recording updates the full recording context, including recordings that you'd set with /load-recordings
GET /recording returns the full recording context, including recordings

File: index.js [hunk 9]
```diff
@@ -168,7 +213,80 @@ app.all('*', (req, res) => {
...
+        res.send(body);
+      }
+
+      if (!recordings[hash]) {
+        recordings[hash] = [];
+      }
+      recordings[hash].push({
+        body: typeof body === 'string' ? body : JSON.stringify(body),
+        status,
+        hashData: {
```
Comment (nickredmark) [inline index.js:268]: I think that to call it "request" would be more informative. Would make data in recordings folder more legible.

File: index.js [hunk 10]
```diff
@@ -168,7 +213,80 @@ app.all('*', (req, res) => {
...
+      } else {
+        body = await proxyRes.text();
+        res.send(body);
+      }
+
+      if (!recordings[hash]) {
+        recordings[hash] = [];
+      }
+      recordings[hash].push({
+        body: typeof body === 'string' ? body : JSON.stringify(body),
```
Comment (nickredmark) [inline index.js:266]: maybe don't stringify

**Pull request merged**